### PR TITLE
Remove mutability from parser/tokenizer APIs.

### PR DESCRIPTION
**Pull request by jdm**
This is incompatible with a re-entrant parsing algorithm which is required by the web platform's document.write API.

### CODE REVIEW

**Comment by jdm** [comment]
FYI @Taym95

File: html5ever/examples/noop-tree-builder.rs [hunk 1]
```diff
@@ -60,54 +61,59 @@ impl TreeSink for Sink {
...
     }
 
-    fn elem_name(&self, target: &usize) -> ExpandedName {
-        self.names.get(target).expect("not an element").expanded()
+    fn elem_name(&self, _target: &usize) -> ExpandedName {
+        //XXX(jdm)
+        //let names = self.names.borrow();
+        //Ref::map(names, |names| names.get(target).expect("not an element").expanded())
+        //self.names.get(target).expect("not an element").expanded()
+        todo!()
```
Comment (jdm) [inline html5ever/examples/noop-tree-builder.rs:69]: I still need to figure out what to do about this code.

File: html5ever/examples/print-tree-actions.rs [hunk 2]
```diff
@@ -41,69 +42,71 @@ impl TreeSink for Sink {
...
     fn same_node(&self, x: &usize, y: &usize) -> bool {
         x == y
     }
 
-    fn elem_name(&self, target: &usize) -> ExpandedName {
-        self.names.get(target).expect("not an element").expanded()
+    fn elem_name(&self, _target: &usize) -> ExpandedName {
+        //XXX(jdm)
+        //self.names.borrow().get(target).cloned().expect("not an element").expanded()
+        todo!()
```
Comment (jdm) [inline html5ever/examples/print-tree-actions.rs:73]: Same here.

File: html5ever/src/macros.rs [hunk 3]
```diff
@@ -9,9 +9,9 @@
 
 macro_rules! unwrap_or_else {
     ($opt:expr, $else_block:block) => {
-        match $opt {
-            None => $else_block,
-            Some(x) => x,
+        {
+            let Some(x) = $opt else { $else_block };
+            x
```
Comment (jdm) [inline html5ever/src/macros.rs:14]: This change is necessary to minimize the scope of refcell borrows that are passed as arguments to the macro.

**Comment by jdm** [comment]
> > FYI @Taym95
> 
> Thanks, Should I close the other PR? lets me knew is you need help here in clean up.

We can still merge the other PR, but we should remove all uses of `&mut BufferQueue` from the codebase in it (and uses like `&mut input`).

**Comment by jdm** [comment]
Ok, the code's cleaned up and commits are split into:
1. mechanical changes - lots of removing `&mut self` from methods and adding Cell/RefCell
2. new APIs needed by https://github.com/servo/servo/pull/32820
3. functional changes - the TreeSink API for obtaining an element's name relies on borrowed data in ways that do not play nicely with naive usage of RefCell to store the element name data ([ExpandedName docs)](https://docs.rs/markup5ever/latest/markup5ever/interface/struct.ExpandedName.html)
4-7. getting CI to pass

**Comment by jdm** [comment]
@Taym95 Would you like to review this?

**Review by Taym95 — APPROVED** [review]
I've reviewed all the changes, and everything looks good to me!

**Review by mrobinson — APPROVED** [review]

**Pull request merged**